File: python/monte_carlo.py
import sys
import time
import random
from multiprocessing import Pool

NUM_PROCESSES = 4

def count_inside(n):
    inside = 0
    for _ in range(n):
        x = random.random()
        y = random.random()
        if x*x + y*y <= 1.0:
            inside += 1
    return inside

def main():
    if len(sys.argv) < 2:
        print("Usage: python monte_carlo.py <num_points>")
        sys.exit(1)

    total = int(sys.argv[1])
    per_proc = total // NUM_PROCESSES

    t_start = time.perf_counter()

    with Pool(NUM_PROCESSES) as pool:
        results = pool.map(count_inside, [per_proc] * NUM_PROCESSES)

    t_end = time.perf_counter()

    inside = sum(results)
    pi = 4.0 * inside / total
    # sanity check
    print(f"pi ~ {pi:.5f}", file=sys.stderr)
    print(f"{t_end - t_start:.6f}")

if __name__ == "__main__":
    main()

File: c/monte_carlo.c
#include <stdio.h>
#include <stdlib.h>
#include <pthread.h>
#include <time.h>

#define NUM_THREADS 4

long long pts_per_thread;
long long counts[NUM_THREADS];

void* run(void *arg) {
    int id = *(int *)arg;
    unsigned int seed = time(NULL) ^ (id * 1234567);
    long long inside = 0;

    for (long long i = 0; i < pts_per_thread; i++) {
        double x = (double)rand_r(&seed) / RAND_MAX;
        double y = (double)rand_r(&seed) / RAND_MAX;
        if (x*x + y*y <= 1.0)
            inside++;
    }

    counts[id] = inside;
    return NULL;
}

int main(int argc, char *argv[]) {
    if (argc < 2) {
        fprintf(stderr, "Usage: %s <num_points>\n", argv[0]);
        return 1;
    }

    long long total = atoll(argv[1]);
    pts_per_thread = total / NUM_THREADS;

    pthread_t threads[NUM_THREADS];
    int ids[NUM_THREADS];

    struct timespec start, end;
    clock_gettime(CLOCK_MONOTONIC, &start);

    for (int t = 0; t < NUM_THREADS; t++) {
        ids[t] = t;
        pthread_create(&threads[t], NULL, run, &ids[t]);
    }
    for (int t = 0; t < NUM_THREADS; t++)
        pthread_join(threads[t], NULL);

    clock_gettime(CLOCK_MONOTONIC, &end);

    long long inside = 0;
    for (int t = 0; t < NUM_THREADS; t++)
        inside += counts[t];

    // just to sanity check the estimate looks right
    fprintf(stderr, "pi ~ %.5f\n", 4.0 * inside / total);

    double elapsed = (end.tv_sec - start.tv_sec) + (end.tv_nsec - start.tv_nsec) / 1e9;
    printf("%.6f\n", elapsed);
    return 0;
}
